async_producer.go

```diff
@@ -218,7 +218,16 @@ func (p *asyncProducer) topicDispatcher() {
 
 		if msg.flags&shutdown != 0 {
```

Comment:
> It would mean this goroutine would need a select instead of a nice simple range... not sure if that really ends up being any cleaner, I'll mock it quickly to see.

Really no simpler, because once the `closing` chan is closed then we have to go to a different loop somewhere or we'll spin forever because a closed channel is always ready for reading (and we still have to process things coming in on the input channel for a while longer).